### ISSUE TITLE
Ch. 15 Code Bug: Some elements of 'result' do not have class `tune_results`

### ISSUE THREAD

**Issue by gserapio** [tidymodels/TMwR#133]
## The problem

The tutorial code in Ch. 15 is returning errors that prevent readers from viewing the tuning results:
```
Error: Some elements of 'result' do not have class `tune_results`.
```
Perhaps these errors are related to `tune` and `workflowsets`?

In addition to creating a reprex below (that uses the default supplied data), I've ran the same code on my own project data and the errors persist. Since the error comes up as `SVM_radial` tunes, I also tried running the code with all of the models listed in the `normalized` workflow set removed, to no avail.

Lastly, the following warnings when tuning the `xgboost` model in the tutorial seem a bit strange:
```
A correlation computation is required, but `estimate` is constant and has 0 standard deviation, resulting in a divide by 0 error. `NA` will be returned.
```
The warnings are mentioned in tidymodels/tune#37 and tidymodels/tune#116. I don't quite understand these warnings because I've successfully tuned my own `xgboost` models on similar data without encountering them. Would they be suppressed if `rsq` wasn't calculated?

Overall, I think this is quite an important and very useful chapter! Thank you for writing it!

## Reproducible example

```r
library(tidymodels)
data(concrete, package = "modeldata")

concrete <- 
  concrete %>% 
  group_by(cement, blast_furnace_slag, fly_ash, water, superplasticizer, 
           coarse_aggregate, fine_aggregate, age) %>% 
  summarize(compressive_strength = mean(compressive_strength),
            .groups = "drop")

set.seed(1501)
concrete_split <- initial_split(concrete, strata = compressive_strength)
concrete_train <- training(concrete_split)
concrete_test  <- testing(concrete_split)

set.seed(1502)
concrete_folds <- 
  vfold_cv(concrete_train, strata = compressive_strength, repeats = 5)

normalized_rec <- 
  recipe(compressive_strength ~ ., data = concrete_train) %>% 
  step_normalize(all_predictors()) 

poly_recipe <- 
  normalized_rec %>% 
  step_poly(all_predictors()) %>% 
  step_interact(~ all_predictors():all_predictors())

library(rules)
#> 
#> Attaching package: 'rules'
#> The following object is masked from 'package:dials':
#> 
#>     max_rules
library(baguette)

linear_reg_spec <- 
  linear_reg(penalty = tune(), mixture = tune()) %>% 
  set_engine("glmnet")

nnet_spec <- 
  mlp(hidden_units = tune(), penalty = tune(), epochs = tune()) %>% 
  set_engine("nnet", MaxNWts = 2600) %>% 
  set_mode("regression")

mars_spec <- 
  mars(prod_degree = tune()) %>%  #<- use GCV to choose terms
  set_engine("earth") %>% 
  set_mode("regression")

svm_r_spec <- 
  svm_rbf(cost = tune(), rbf_sigma = tune()) %>% 
  set_engine("kernlab") %>% 
  set_mode("regression")

svm_p_spec <- 
  svm_poly(cost = tune(), degree = tune()) %>% 
  set_engine("kernlab") %>% 
  set_mode("regression")

knn_spec <- 
  nearest_neighbor(neighbors = tune(), dist_power = tune(), weight_func = tune()) %>% 
  set_engine("kknn") %>% 
  set_mode("regression")

cart_spec <- 
  decision_tree(cost_complexity = tune(), min_n = tune()) %>% 
  set_engine("rpart") %>% 
  set_mode("regression")

bag_cart_spec <- 
  bag_tree() %>% 
  set_engine("rpart", times = 50L) %>% 
  set_mode("regression")

rf_spec <- 
  rand_forest(mtry = tune(), min_n = tune(), trees = 1000) %>% 
  set_engine("ranger") %>% 
  set_mode("regression")

xgb_spec <- 
  boost_tree(tree_depth = tune(), learn_rate = tune(), loss_reduction = tune(), 
             min_n = tune(), sample_size = tune(), trees = tune()) %>% 
  set_engine("xgboost") %>% 
  set_mode("regression")

cubist_spec <- 
  cubist_rules(committees = tune(), neighbors = tune()) %>% 
  set_engine("Cubist") 

nnet_param <- 
  nnet_spec %>% 
  parameters() %>% 
  update(hidden_units = hidden_units(c(1, 27)))

library(workflowsets)

normalized <- 
  workflow_set(
    preproc = list(normalized = normalized_rec), 
    models = list(SVM_radial = svm_r_spec, SVM_poly = svm_p_spec, 
                  KNN = knn_spec, neural_network = nnet_spec)
  )

normalized <- 
  normalized %>% 
  option_add(param = nnet_param, id = "normalized_neural_network")

model_vars <- 
  workflow_variables(outcomes = compressive_strength, 
                     predictors = everything())

no_pre_proc <- 
  workflow_set(
    preproc = list(simple = model_vars), 
    models = list(MARS = mars_spec, CART = cart_spec, CART_bagged = bag_cart_spec,
                  RF = rf_spec, boosting = xgb_spec, Cubist = cubist_spec)
  )

with_features <- 
  workflow_set(
    preproc = list(full_quad = poly_recipe), 
    models = list(linear_reg = linear_reg_spec, KNN = knn_spec)
  )

all_workflows <- 
  bind_rows(no_pre_proc, normalized, with_features) %>% 
  # Make the workflow ID's a little more simple: 
  mutate(wflow_id = gsub("(simple_)|(normalized_)", "", wflow_id))

grid_ctrl <-
  control_grid(
    save_pred = TRUE,
    parallel_over = "everything",
    save_workflow = TRUE
  )

full_results_time <- 
  system.time(
    grid_results <- 
      all_workflows %>% 
      workflow_map(seed = 1503, resamples = concrete_folds, grid = 25, 
                   control = grid_ctrl, verbose = TRUE)
  )
#> i  1 of 12 tuning:     MARS
#> ✓  1 of 12 tuning:     MARS (26s)
#> i  2 of 12 tuning:     CART
#> ✓  2 of 12 tuning:     CART (5m 23s)
#> i    No tuning parameters. `fit_resamples()` will be attempted
#> i  3 of 12 resampling: CART_bagged
#> ✓  3 of 12 resampling: CART_bagged (7m 17.1s)
#> i  4 of 12 tuning:     RF
#> i Creating pre-processing data to finalize unknown parameter: mtry
#> ✓  4 of 12 tuning:     RF (11m 31.3s)
#> i  5 of 12 tuning:     boosting
#> ! Fold01, Repeat1: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold01, Repeat1: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold02, Repeat1: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold02, Repeat1: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold03, Repeat1: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold03, Repeat1: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold04, Repeat1: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold04, Repeat1: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold05, Repeat1: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold05, Repeat1: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold06, Repeat1: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold06, Repeat1: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold07, Repeat1: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold07, Repeat1: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold08, Repeat1: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold08, Repeat1: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold09, Repeat1: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold09, Repeat1: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold10, Repeat1: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold10, Repeat1: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold01, Repeat2: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold01, Repeat2: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold02, Repeat2: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold02, Repeat2: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold03, Repeat2: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold03, Repeat2: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold04, Repeat2: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold04, Repeat2: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold05, Repeat2: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold05, Repeat2: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold06, Repeat2: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold06, Repeat2: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold07, Repeat2: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold07, Repeat2: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold08, Repeat2: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold08, Repeat2: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold09, Repeat2: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold09, Repeat2: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold10, Repeat2: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold10, Repeat2: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold01, Repeat3: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold01, Repeat3: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold02, Repeat3: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold02, Repeat3: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold03, Repeat3: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold03, Repeat3: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold04, Repeat3: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold04, Repeat3: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold05, Repeat3: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold05, Repeat3: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold06, Repeat3: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold06, Repeat3: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold07, Repeat3: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold07, Repeat3: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold08, Repeat3: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold08, Repeat3: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold09, Repeat3: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold09, Repeat3: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold10, Repeat3: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold10, Repeat3: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold01, Repeat4: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold01, Repeat4: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold02, Repeat4: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold02, Repeat4: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold03, Repeat4: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold03, Repeat4: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold04, Repeat4: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold04, Repeat4: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold05, Repeat4: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold05, Repeat4: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold06, Repeat4: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold06, Repeat4: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold07, Repeat4: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold07, Repeat4: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold08, Repeat4: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold08, Repeat4: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold09, Repeat4: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold09, Repeat4: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold10, Repeat4: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold10, Repeat4: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold01, Repeat5: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold01, Repeat5: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold02, Repeat5: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold02, Repeat5: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold03, Repeat5: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold03, Repeat5: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold04, Repeat5: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold04, Repeat5: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold05, Repeat5: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold05, Repeat5: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold06, Repeat5: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold06, Repeat5: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold07, Repeat5: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold07, Repeat5: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold08, Repeat5: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold08, Repeat5: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold09, Repeat5: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold09, Repeat5: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold10, Repeat5: internal: A correlation computation is required, but `estimate` is const...
#> ! Fold10, Repeat5: internal: A correlation computation is required, but `estimate` is const...
#> ✓  5 of 12 tuning:     boosting (19m 36.3s)
#> i  6 of 12 tuning:     Cubist
#> ✓  6 of 12 tuning:     Cubist (19m 30.8s)
#> i  7 of 12 tuning:     SVM_radial
#> Error: Some elements of 'result' do not have class `tune_results`.
#> Execution stepped; returning current results
#> Error: Some elements of 'result' do not have class `tune_results`.
#> Timing stopped at: 3821 40.79 3826
```

<sup>Created on 2021-04-14 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

Session info:
```
R version 4.0.3 (2020-10-10)
Platform: x86_64-apple-darwin17.0 (64-bit)
Running under: macOS Big Sur 10.16

Matrix products: default
LAPACK: /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] reprex_2.0.0       Cubist_0.2.3       lattice_0.20-41    xgboost_1.3.2.1    ranger_0.12.1      rpart_4.1-15      
 [7] earth_5.3.0        plotmo_3.6.0       TeachingDemos_2.12 plotrix_3.8-1      Formula_1.2-4      vctrs_0.3.7       
[13] rlang_0.4.10       workflowsets_0.0.1 baguette_0.1.0     rules_0.1.1        yardstick_0.0.8    workflows_0.2.2   
[19] tune_0.1.3         tidyr_1.1.3        tibble_3.1.0       rsample_0.0.9      recipes_0.1.15     purrr_0.3.4       
[25] parsnip_0.1.5      modeldata_0.1.0    infer_0.5.4        ggplot2_3.3.3      dplyr_1.0.5        dials_0.0.9       
[31] scales_1.1.1       broom_0.7.6        tidymodels_0.1.2  

loaded via a namespace (and not attached):
 [1] colorspace_2.0-0   ellipsis_0.3.1     class_7.3-18       fs_1.5.0           rstudioapi_0.13    listenv_0.8.0     
 [7] furrr_0.2.2        prodlim_2019.11.13 fansi_0.4.2        mvtnorm_1.1-1      lubridate_1.7.10   codetools_0.2-18  
[13] splines_4.0.3      libcoin_1.0-8      knitr_1.31         pROC_1.17.0.1      clipr_0.7.1        compiler_4.0.3    
[19] backports_1.2.1    assertthat_0.2.1   Matrix_1.3-2       cli_2.4.0          htmltools_0.5.1.1  prettyunits_1.1.1 
[25] tools_4.0.3        partykit_1.2-13    gtable_0.3.0       glue_1.4.2         reshape2_1.4.4     Rcpp_1.0.6        
[31] styler_1.4.1       DiceDesign_1.9     iterators_1.0.13   timeDate_3043.102  inum_1.0-4         gower_0.2.2       
[37] xfun_0.22          stringr_1.4.0      globals_0.14.0     ps_1.6.0           lifecycle_1.0.0    future_1.21.0     
[43] MASS_7.3-53.1      ipred_0.9-11       parallel_4.0.3     yaml_2.2.1         C50_0.1.3.1        stringi_1.5.3     
[49] highr_0.8          foreach_1.5.1      butcher_0.1.4      lhs_1.1.1          hardhat_0.1.5      lava_1.6.9        
[55] pkgconfig_2.0.3    evaluate_0.14      tidyselect_1.1.0   processx_3.5.1     parallelly_1.24.0  plyr_1.8.6        
[61] magrittr_2.0.1     R6_2.5.0           generics_0.1.0     DBI_1.1.1          pillar_1.6.0       withr_2.4.1       
[67] survival_3.2-10    nnet_7.3-15        crayon_1.4.1       utf8_1.2.1         rmarkdown_2.7      usethis_2.0.1     
[73] grid_4.0.3         data.table_1.14.0  callr_3.6.0        digest_0.6.27      GPfit_1.0-8        munsell_0.5.0  
```

**Comment by juliasilge** [comment]
I believe you mean [Ch 15](https://www.tmwr.org/workflow-sets.html), not Ch 11, correct?

Check out the fix explained in tidymodels/workflowsets#41 and try installing the current development version of workflowsets for that:

```r
devtools::install_github("tidymodels/workflowsets")
```

We believe that should solve the problem.

**Comment by gserapio** [comment]
> I believe you mean [Ch 15](https://www.tmwr.org/workflow-sets.html), not Ch 11, correct?
> 
> Check out the fix explained in [tidymodels/workflowsets#41](https://github.com/tidymodels/workflowsets/issues/41) and try installing the current development version of workflowsets for that:
> 
> ```r
> devtools::install_github("tidymodels/workflowsets")
> ```
> 
> We believe that should solve the problem.

Whoops, yes Ch. 15 (edited to correct that). Sorry I overlooked that issue! Thanks so much, I will try this!

**Comment by juliasilge** [comment]
Let us know if you have further problems!
